gemma.cc

```diff
@@ -721,50 +726,53 @@ HWY_EXPORT(GetCompressedWeightsT);
 HWY_EXPORT(Generate2B);
 HWY_EXPORT(Generate7B);
 
-KVCache CreateKVCache(size_t size_cache_pos, size_t seq_len) {
-  KVCache kv_cache = {};
-  kv_cache.key_cache = hwy::AllocateAligned<float>(seq_len * size_cache_pos);
-  kv_cache.value_cache = hwy::AllocateAligned<float>(seq_len * size_cache_pos);
-  return kv_cache;
-}
-
 template <class Config>
 GemmaImpl<Config>::GemmaImpl(const LoaderArgs& args, hwy::ThreadPool& pool)
     : compressed_weights(
           HWY_DYNAMIC_DISPATCH(GetCompressedWeightsT)(args, pool)),
       prefill(hwy::MakeUniqueAligned<Activations<Config, kPrefillBatchSize>>()),
-      state(hwy::MakeUniqueAligned<Activations<Config, 1>>()),
-      kv_cache(
-          CreateKVCache(Config::kLayers * Config::kKVHeads * Config::kQKVDim,
-                        Config::kSeqLen)) {
+      state(hwy::MakeUniqueAligned<Activations<Config, 1>>()) {
   PROFILER_ZONE("Startup.tokenizer");
 
   HWY_ASSERT(tokenizer.Load(args.tokenizer.path).ok());
 }
 
+template <class Config>
+KVCache GemmaImpl<Config>::CreateKVCache() const {
```

Comment:
In the original version, the KV cache is created in the constructor of class `GemmaImpl`, the size of the cache depends on the structure of the model defined in `Config`. For processing multiple ongoing conversations, each conversation requires a separate KV cache, so we need to create a KV cache before each conversation starts, and its size must match the model structure we used. In my opinion, the intuitive thing to do is to create it via the instance of Gemma. However, the model structure is not yet exposed to users of the high-level API, so I need to use the dynamic dispatch of `GemmaInterface` to implement it. That's why I made it a method instead of a free function.

Reply:
That's a fair assessment given the current state of the implementation... although it's probably not that obvious, the Config structs are used is more like a configuration for the entire runtime, which consists of mostly model parameters but also folds in kSeqLen and kTopK which are not really model parameters.

This was an okay design for the initial "vanilla" interactive use case. As we flesh out the library usage, it's easy to imagine an application where the same model is interacting with different KV caches of different sizes and with multiple samplers that do different things (eg imagine a system with different scratchpads for code / reasoning / RAG with different properties, or a multi-agent system with a shared set of weights etc).

Somewhat orthogonal to this, we want to simplify the management of KV cache limits - having all 3 of kSeqLen, max_len, and max_generathion_length is more complicated than I'd like. One solution is to have the sizing be more dynamic, potentially allocating more as needed up to some limit. There's some pointer arithmetic things that would need to be updated to do this. 

So the general direction is probably to break up the coupling between runtime and model to allow for more flexible composability in the library, at least for the lower level *Impl layers. We maintain some of the higher-level / less-verbose interfaces for common usages but hoist them to sit above the core implementation layer as factory functions or possibly convenience wrapper types - sitting at the boundary between the "frontends" and "models" layers (rather than having them baked into the core model code).

You're right CreateKVCache is somewhat coupled to the model parameters today. If it's possible to do without adding too much complexity, I'd like to make changes in this direction of proactively decoupling in this direction, rather than having to peel things back later. If you're up for making the change that would be helpful, but I'd also be okay with merging this as is for now and refining this as we refactor the model/library interfaces.

Reply:
Thanks for the detailed response, I probably understand what you are considering.

How about resetting to the original `CreateKVCache` function but exposing it in `gemma.h`? Maybe we need to add an auxiliary template function to call it like the original constructor of `GemmaImpl` and determine which `Config` to use based on `ModelType` method of `LoaderArgs`. Then `GemmaInterface` will not change, which can not only solve the current problem but also be beneficial for future refining.

Of course, this is only a workaround solution. If it's acceptable, I'll do it tomorrow.

Reply:
Yes we can try exposing it in `gemma.h`. 

The direction I'm moving is gradually moving to factory functions to decouple core types from how their state is populated. By making GemmaImpl decoupled from disk loading concerns, that should make it easier for different frontends wit different needs (eg reading from disk vs streaming) to do whatever they need to without running into roadblocks. 

I've made some updates to the library interfaces in this PR https://github.com/google/gemma.cpp/pull/82 (in the examples branch). It's not finished + has some additional polishing to do, especially at the top level `Gemma` constructor, but hopefully that gives a more concrete idea of the general direction.

Might consider targeting this PR to that examples/ branch if it doesn't add too much complexity.